### PR TITLE
Add Cookies

### PR DESCRIPTION
**Pull request by Commandcracker**
## Description
Cookies work, but there is nothing to really use them because pumpkin has no plugin support yet.

## Testing
in command
```rs
if let CommandSender::Player(p) = sender {
        let v = &vec![255, 0, 255];
            p.client
                .send_packet(&CStoreCookie::new(
                    &String::from("funnyplugin:hello"),
                    &VarInt(v.len() as i32),
                    v,
                ))
                .await;
        }
        if let CommandSender::Player(p) = sender {
            p.client
                .send_packet(&CCookieRequest::new(&String::from("funnyplugin:hello")))
                .await;
        }
}
```
2024-11-22 21:05:07 INFO  [pumpkin::client::player_packet@tokio-runtime-worker] Player (Commandcracker): executed command /version
2024-11-22 21:05:07 INFO  [pumpkin::client::player_packet@tokio-runtime-worker] Received cookie_response[play]: key: "funnyplugin:hello", has_payload: "true", payload_length: "3"

## Checklist
Things need to be done before this Pull Request can be merged.

- [x] Code is well-formatted and adheres to project style guidelines: `cargo fmt`
- [x] Code does not produce any clippy warnings `cargo clippy`

### CODE REVIEW

File: pumpkin-protocol/src/client/config/c_store_cookie.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+use pumpkin_macros::client_packet;
+
+use crate::{Identifier, VarInt};
+
+#[derive(serde::Serialize)]
+#[client_packet("config:store_cookie")]
+/// Stores some arbitrary data on the client, which persists between server transfers.
+/// The Notchian client only accepts cookies of up to 5 kiB in size.
```
Comment (DataM0del) [inline pumpkin-protocol/src/client/config/c_store_cookie.rs:8]:
```suggestion
/// The Notchian (vanilla) client only accepts cookies of up to 5 kiB in size.
```
Most people should understand what the "Notchian" client means, but... Why not?

File: pumpkin-protocol/src/server/config/s_cookie_response.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+use pumpkin_macros::server_packet;
+
+use crate::{Identifier, VarInt};
+
+#[derive(serde::Deserialize)]
+#[server_packet("config:cookie_response")]
+/// Response to a Cookie Request (configuration) from the server.
+/// The Notchian server only accepts responses of up to 5 kiB in size.
```
Comment (DataM0del) [inline pumpkin-protocol/src/server/config/s_cookie_response.rs:8]:
```suggestion
/// The Notchian (vanilla) server only accepts responses of up to 5 kiB in size.
```
Same reason as previously.

File: pumpkin-protocol/src/server/play/s_cookie_response.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+use pumpkin_macros::server_packet;
+use crate::{Identifier, VarInt};
+
+#[derive(serde::Deserialize, Debug)]
+#[server_packet("play:cookie_response")]
+/// Response to a Cookie Request (play) from the server.
+/// The Notchian server only accepts responses of up to 5 kiB in size.
```
Comment (DataM0del) [inline pumpkin-protocol/src/server/play/s_cookie_response.rs:7]:
```suggestion
/// The Notchian (vanilla) server only accepts responses of up to 5 kiB in size.
```
Same reason as previously.

**Review by DataM0del — COMMENTED** [review]
CO>>**KK**<<IES PROTOCOL :skull: (typo moment)
also, not sure if you can trick the server and send a payload longer than 5 kilobytes (kB) and setting a fake length. Would that just crash the server or something? Will look after this review to see.

**Comment by DataM0del** [comment]
Nothing succeeded :skull:

**Comment by DataM0del** [comment]
Why......

**Comment by DataM0del** [comment]
Oh... perfect!
image

**Comment by Commandcracker** [comment]
> CO>>**KK**<<IES PROTOCOL 💀 (typo moment) ...

Didn't want to waste my time trying to change a commit message before pushing.

> ...also, not sure if you can trick the server and send a payload longer than 5 kilobytes (kB) and setting a fake length. Would that just crash the server or something? Will look after this review to see.

Should try it, I just added to protocol for now.

**Comment by Commandcracker** [comment]
> Nothing succeeded 💀
> Oh... perfect!

I can also see the GH actions ._.
Did you even read this ?
very **WIP** just made this **draft**, so everyone can **see that in working on it**

**Review by DataM0del — APPROVED** [review]

File: pumpkin-protocol/src/client/config/c_store_cookie.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+use crate::{Identifier, VarInt};
+
+#[derive(serde::Serialize)]
+#[client_packet("config:store_cookie")]
+/// Stores some arbitrary data on the client, which persists between server transfers.
+/// The Notchian (vanilla) client only accepts cookies of up to 5 kiB in size.
+pub struct CStoreCookie<'a> {
+    key: &'a Identifier,
+    payload_length: &'a VarInt,
+    payload: &'a Vec<u8>, // 5120,
```
Comment (Snowiiii) [inline pumpkin-protocol/src/client/config/c_store_cookie.rs:12]: use `&[u8]` not a Vec

File: pumpkin-protocol/src/client/config/c_store_cookie.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+impl<'a> CStoreCookie<'a> {
+    pub fn new(key: &'a Identifier, payload_length: &'a VarInt, payload: &'a Vec<u8>) -> Self {
+        Self {
+            key,
+            payload_length,
```
Comment (Snowiiii) [inline pumpkin-protocol/src/client/config/c_store_cookie.rs:19]: Don't require the payload length. Just get it your own 
```
payload_length: payload.len()
```

File: pumpkin-protocol/src/client/play/c_store_cookie.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    key: &'a Identifier,
+    payload_length: &'a VarInt,
+    payload: &'a Vec<u8>, // 5120,
+}
+
+impl<'a> CStoreCookie<'a> {
+    pub fn new(key: &'a Identifier, payload_length: &'a VarInt, payload: &'a Vec<u8>) -> Self {
+        Self {
+            key,
+            payload_length,
```
Comment (Snowiiii) [inline pumpkin-protocol/src/client/play/c_store_cookie.rs:19]: Here the same

File: pumpkin/src/client/client_packet.rs [hunk 7]
```diff
@@ -251,6 +252,17 @@ impl Client {
         Err(AuthError::MissingAuthClient)
     }
 
+    #[expect(clippy::unused_async)]
+    pub async fn handle_login_cookie_response(&self, packet: SLCookieResponse) {
```
Comment (Snowiiii) [inline pumpkin/src/client/client_packet.rs:256]: Just make it not async

File: pumpkin/src/client/client_packet.rs [hunk 8]
```diff
@@ -342,6 +354,17 @@ impl Client {
         }
     }
 
+    #[expect(clippy::unused_async)]
+    pub async fn handle_config_cookie_response(&self, packet: SCCookieResponse) {
```
Comment (Snowiiii) [inline pumpkin/src/client/client_packet.rs:358]: Just make it not async

**Comment by Snowiiii** [comment]
Looks good to me :+1: 
Thanks @Commandcracker :heart:

**Pull request merged**